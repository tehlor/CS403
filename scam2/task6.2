(println (powerSet '(y w e b u r)))
