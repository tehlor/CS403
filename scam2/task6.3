(println (powerSet '(z k t v u)))
