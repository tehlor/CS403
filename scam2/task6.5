(println (powerSet '(u r i p e w b)))
