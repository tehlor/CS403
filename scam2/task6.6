(println (powerSet '(l w k e)))
