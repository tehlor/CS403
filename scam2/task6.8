(println (powerSet '(u h g q a p t y l)))
