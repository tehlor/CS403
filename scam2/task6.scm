(define (main)
    (setNilDisplay 'nil)
    (setPort (open (getElement ScamArgs 1) 'read))
    (readArgs)
)

(define (readArgs)
    (define env this)
    (define (iter expr)
        (if (not (eof?))
            (begin (eval expr env) 
                (iter (readExpr))
            )
        )
    )
    (iter (readExpr))
)
;; let psetRest be all obtained subsets (ps (cdr set)) so we dont have to call it twice below 
;; append the list of all current subsets to the new subsets obtained by adding (car set) to all current subsets
;; cons car to all subsets of pset rest to obtain the new subsets
(define (F e T)
  (map
    (lambda (X)
      (cons e X)
      )
    T
    )
  )

(define (powerSet S)
  (cond
    ((null? S) '(()) )
    (else 
      (let ( (T (powerSet (cdr S))) (e (car S)) )
        (merge
          T
          (F e T)
          )
        )
      )
    )
  )

(define (merge a b)
  (define (help f s)
    (cond 
      ((null? f) s)
      ((null? s) f)
      (else
            (if (< (length (car f)) (length (car s)))
                (cons (car f) (help (cdr f) s))
                (cons (car s) (help f (cdr s)))
            )
        )
      )
    )
  (help a b)
)

