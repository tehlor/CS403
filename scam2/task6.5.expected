(nil (u) (r) (i) (p) (e) (w) (b) (u r) (u i) (u p) (u e) (u w) (u b) (r i) (r p) (r e) (r w) (r b) (i p) (i e) (i w) (i b) (p e) (p w) (p b) (e w) (e b) (w b) (u r i) (u r p) (u r e) (u r w) (u r b) (u i p) (u i e) (u i w) (u i b) (u p e) (u p w) (u p b) (u e w) (u e b) (u w b) (r i p) (r i e) (r i w) (r i b) (r p e) (r p w) (r p b) (r e w) (r e b) (r w b) (i p e) (i p w) (i p b) (i e w) (i e b) (i w b) (p e w) (p e b) (p w b) (e w b) (u r i p) (u r i e) (u r i w) (u r i b) (u r p e) (u r p w) (u r p b) (u r e w) (u r e b) (u r w b) (u i p e) (u i p w) (u i p b) (u i e w) (u i e b) (u i w b) (u p e w) (u p e b) (u p w b) (u e w b) (r i p e) (r i p w) (r i p b) (r i e w) (r i e b) (r i w b) (r p e w) (r p e b) (r p w b) (r e w b) (i p e w) (i p e b) (i p w b) (i e w b) (p e w b) (u r i p e) (u r i p w) (u r i p b) (u r i e w) (u r i e b) (u r i w b) (u r p e w) (u r p e b) (u r p w b) (u r e w b) (u i p e w) (u i p e b) (u i p w b) (u i e w b) (u p e w b) (r i p e w) (r i p e b) (r i p w b) (r i e w b) (r p e w b) (i p e w b) (u r i p e w) (u r i p e b) (u r i p w b) (u r i e w b) (u r p e w b) (u i p e w b) (r i p e w b) (u r i p e w b))
